core/translate/mod.rs

```diff
@@ -357,42 +337,102 @@ fn translate_pragma(
     Ok(program.build(database_header, connection))
 }
 
-fn update_pragma(name: &str, value: i64, header: Rc<RefCell<DatabaseHeader>>, pager: Rc<Pager>) {
-    match name {
-        "cache_size" => {
-            let mut cache_size_unformatted = value;
-            let mut cache_size = if cache_size_unformatted < 0 {
-                let kb = cache_size_unformatted.abs() * 1024;
-                kb / 512 // assume 512 page size for now
-            } else {
-                value
-            } as usize;
-            if cache_size < MIN_PAGE_CACHE_SIZE {
-                // update both in memory and stored disk value
-                cache_size = MIN_PAGE_CACHE_SIZE;
-                cache_size_unformatted = MIN_PAGE_CACHE_SIZE as i64;
-            }
-
-            // update in-memory header
-            header.borrow_mut().default_cache_size = cache_size_unformatted
-                .try_into()
-                .unwrap_or_else(|_| panic!("invalid value, too big for a i32 {}", value));
+fn update_pragma(
+    name: &str,
+    value: ast::Expr,
+    header: Rc<RefCell<DatabaseHeader>>,
+    pager: Rc<Pager>,
+    program: &mut ProgramBuilder,
+) {
+    let pragma = PragmaName::from_str(name).expect("provided pragma not valid");
+    match pragma {
+        PragmaName::CacheSize => {
+            let cache_size = match value {
+                ast::Expr::Literal(ast::Literal::Numeric(numeric_value)) => {
+                    numeric_value.parse::<i64>().unwrap()
+                }
+                ast::Expr::Unary(ast::UnaryOperator::Negative, expr) => match *expr {
+                    ast::Expr::Literal(ast::Literal::Numeric(numeric_value)) => {
+                        -numeric_value.parse::<i64>().unwrap()
+                    }
+                    _ => 0,
+                },
+                _ => 0,
```

Comment:
I think instead of returning 0 it should be either a todo or a parse error. We could miss a case that was correct if we do it like this.